File: server/test/fixtures.js
/**
 * Genera archivos .xlsx minimos con el mismo acomodo que los de planta, para
 * que los tests no dependan de los documentos reales de Florence.
 */
import ExcelJS from 'exceljs';
import { COLUMNAS, HOJA_WI } from '../src/ingesta/parametros.js';

/** WI de juguete: { diametro: { columna: mmS } }. */
export async function crearWi(ruta, velocidades) {
  const wb = new ExcelJS.Workbook();
  const ws = wb.addWorksheet(HOJA_WI);

  ws.getCell('A1').value = 'ITW Line Speed\n[mm/s]';
  ws.getCell('A2').value = 'Drawn Wire Rod Ø\n[mm]';
  for (const [columna, { lineas }] of COLUMNAS) ws.getRow(2).getCell(columna).value = lineas.join('/');

  let fila = 3;
  for (const diametro of Object.keys(velocidades).map(Number).sort((a, b) => a - b)) {
    ws.getRow(fila).getCell(1).value = diametro;
    for (const [columna, mmS] of Object.entries(velocidades[diametro])) {
      ws.getRow(fila).getCell(Number(columna)).value = mmS;
    }
    fila++;
  }
  await wb.xlsx.writeFile(ruta);
  return ruta;
}

const ENCABEZADOS = [
  'Work Center',
  'Material Number',
  'Material Description',
  'Important Notes',
  'Order',
  'PO Printed Yes/No',
  'Drawn Yes/No',
  'Operation Quantity (MEINH)',
  'Customer PO',
];

/** Schedule de juguete. Cada renglon: [wc, material, descripcion, notas, orden, kg]. */
export async function crearSchedule(ruta, renglones) {
  const wb = new ExcelJS.Workbook();
  const ws = wb.addWorksheet('Sheet1');

  ws.getCell('A1').value = 'CSW Production Schedule (prueba)';
  ENCABEZADOS.forEach((e, i) => (ws.getRow(2).getCell(i + 1).value = e));

  renglones.forEach(([wc, material, descripcion, notas, orden, kg], i) => {
    const fila = ws.getRow(3 + i);
    fila.getCell(1).value = wc;
    fila.getCell(2).value = material;
    fila.getCell(3).value = descripcion;
    fila.getCell(4).value = notas;
    fila.getCell(5).value = orden;
    fila.getCell(8).value = kg;
  });
  await wb.xlsx.writeFile(ruta);
  return ruta;
}

File: tests/fixtures.py
"""Genera archivos .xlsx minimos con el mismo acomodo que los de planta.

Asi los tests no dependen de los documentos reales de Florence, que no se
versionan.
"""

from __future__ import annotations

from pathlib import Path

from openpyxl import Workbook

from hiloflo.parametros import COLUMNAS, HOJA_PREDETERMINADA


def crear_wi(ruta: Path, velocidades: dict[float, dict[int, float]]) -> Path:
    """Arma un WI de juguete: {diametro: {columna: mm/s}}.

    Replica el acomodo real: encabezados de bloque arriba y la columna A con
    el diametro de alambre estirado.
    """
    wb = Workbook()
    ws = wb.active
    ws.title = HOJA_PREDETERMINADA

    ws.cell(1, 1, "ITW Line Speed\n[mm/s]")
    ws.cell(2, 1, "Drawn Wire Rod Ø\n[mm]")
    for columna, (lineas, *_) in COLUMNAS.items():
        ws.cell(2, columna, "/".join(lineas))

    fila = 3
    for diametro, por_columna in sorted(velocidades.items()):
        ws.cell(fila, 1, diametro)
        for columna, mm_s in por_columna.items():
            ws.cell(fila, columna, mm_s)
        fila += 1

    wb.save(ruta)
    return ruta


ENCABEZADOS_SCHEDULE = [
    "Work Center",
    "Material Number",
    "Material Description",
    "Important Notes",
    "Order",
    "PO Printed Yes/No",
    "Drawn Yes/No",
    "Operation Quantity (MEINH)",
    "Customer PO",
]


def crear_schedule(ruta: Path, renglones: list[tuple]) -> Path:
    """Arma un schedule de juguete.

    Cada renglon es (work_center, material, descripcion, notas, orden, kg).
    """
    wb = Workbook()
    ws = wb.active
    ws.title = "Sheet1"

    ws.cell(1, 1, "CSW Production Schedule (prueba)")
    for col, encabezado in enumerate(ENCABEZADOS_SCHEDULE, start=1):
        ws.cell(2, col, encabezado)

    for fila, (wc, material, descripcion, notas, orden, kg) in enumerate(
        renglones, start=3
    ):
        ws.cell(fila, 1, wc)
        ws.cell(fila, 2, material)
        ws.cell(fila, 3, descripcion)
        ws.cell(fila, 4, notas)
        ws.cell(fila, 5, orden)
        ws.cell(fila, 8, kg)

    wb.save(ruta)
    return ruta
